core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,210 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Closeable;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ *
+ *
+ */
+// Discussion point. Using java.io.Closeable here so we can inherit AutoCloseable where its available.
+public abstract class Context implements Closeable {
+
+  /**
+   * The logical root context which is 'current' if no other context is bound.
+   */
+  public static final Context ROOT = null; // FIXME
+
+  public interface Listener {
+    public void closed(Context context);
+  }
+
+  /**
+   * Create a {@link Key} with the given name.
+   */
+  public static <T> Key<T> key(String name) {
+    return new Key<T>(name);
+  }
+
+  /**
+   * Create a {@link Key} with the given name and default value.
+   */
+  public static <T> Key<T> keyWithDefault(String name, T defaultValue) {
+    return new Key<T>(name, defaultValue);
+  }
+
+  /**
+   * Retrieve the value for the key for this context, will traverse the parent chain.
+   * @param key
+   * @return
+   */
+  // Discussion point, do we want this or do we make everyone use Key including in the builder
+  // Note that we are not providing any iteration of the values in the context.
+  public abstract <T> T get(Key<T> key);
+
+  /**
+   * Return the current context, will never return null.
+   */
+  // Discussion point. Is choosing to always return ROOT if the thread-local is empty useful?
+  public static Context current() {
+    return null;
+  }
+
+
+  /**
+   * Detach the current context from the thread and restore the context that was previously
+   * attached to the thread as the 'current' context. It is perfectly fine to attach a
+   * context which is already closed.
+   *
+   * @throws java.lang.IllegalStateException if this context is not current.
+   * @return the current context.
```

Comment:
I think you need to specify whether that is before or after the detach. If it is before, then that is 'this' object.

When would you expect to use the returned Context?